procwrap/procwrap.py

```diff
@@ -3,173 +3,208 @@
-        # TODO buffering and sending many to nsq together ?
-        # TODO avoid
+
+        hostname = self.hostname # from basescript
         while poll() is None:
             line = read_stderr_line()
-            # TODO pretty print to stderr ?
-            sys.stderr.write(line)
 
-            # the line may not be valid json object
-            if line.startswith("{") and line.endswith("}\n"):
-                # TODO load the json line, add some other parameters like hostname, uuid, etc
-                # serialize again and send to nsq
-                # TODO check that it is a log
+            try:
+                jline = json.loads(line)
+
+                level = jline.get("level", "debug")
+                event = jline.pop("event", "")
+                # NOTE not using standard log levels because we want this to be transparent
+                self.log._proxy_to_logger(level, event, **jline)
 
-                # NOTE for now just sending it across
-                self.write_to_nsq(line)
+                jline['event'] = event
+                jline['uuid'] = str(uuid1())
+                jline['host'] = hostname
+                # we can get the timestamp from the uuid itself, so no need for timestamp
+                self.write_to_nsq(json.dumps(jline))
 
-            # FIXME is this a bottleneck ? only one log per 0.1 seconds ?
-            # consider threads http://eyalarubas.com/python-subproc-nonblock.html
-            time.sleep(0.1)
+            except ValueError:
+                sys.stderr.write(line)
 
         # read the remaining stderr
         remaining = self.process.stderr.read()
+        # TODO this should be processed as well
         sys.stderr.write(remaining)
         sys.stderr.flush()
 
-        return self.process.returncode
-
-def parse_args():
-    # TODO accept configuration file
-
-    parser = argparse.ArgumentParser(
-        description="process wrapper",
-    )
-
-    parser.add_argument("-p", "--pid",
-        default=ProcessWrapper._PID_FILE,
-        help="write the pid to file if present, default: %(default)s"
-    )
-
-    # TODO should this be the other way around ? always disabled ?
-    parser.add_argument("--disable-nsq", action="store_true",
-        default=ProcessWrapper._DISABLE_NSQ,
-        help="to disable pushing to nsq, default: %(default)s",
-    )
-
-    parser.add_argument("--nsq-log-topic",
-        default=ProcessWrapper._NSQ_LOG_TOPIC,
-        help="nsq topic name for logs, default: %(default)s",
-    )
-
-    parser.add_argument("--nsqd-http-address",
-        default=ProcessWrapper._NSQD_HTTP_ADDRESS,
-        help="nsqd http address, default: %(default)s",
-    )
-
-    parser.add_argument("command", nargs=argparse.REMAINDER,
-        help="the command to run",
-    )
-
-    args = parser.parse_args()
-    if args.pid:
-        args.pid = os.path.abspath(os.path.expanduser(args.pid))
+        sys.exit(self.process.returncode)
+
+    def on_exit(self):
```

Comment:
@prashanthellina I haven't been able to figure this out yet, but whenever I send an interrupt (CTRL + C), it seems to be propagating to the child, but not calling my atexit function.

I was reading through http://bugs.python.org/issue14073 and https://www.thecodingforums.com/threads/question-about-ctrl-d-and-atexit-with-threads.674458/

> Reality is that the way Python behaviour is defined/implemented means that it will wait for non daemonised threads to complete before exiting.

So I decided to daemonize it.